README.md
# The Amarok Linux Welcome screen

![Screenshot of Amarok 3.3](/screencaptures/welcome_screen_amarok_3_3.png?raw=true)


debian/changelog
amarokwelcome (2.5.3-1) stable; urgency=medium

  * New upstream version for Amarok Linux OS
  * update

 -- Amarok Linux <packages@amaroklinux.org>  Wed, 10 Aug 2022 12:54:15 -0300

amarokwelcome (2.5.3) stable; urgency=medium

  * New upstream version for Amarok Linux OS

 -- Amarok Linux <packages@amaroklinux.org>  Tue, 09 Aug 2022 15:43:10 -0300

amarokwelcome (2.5.2+amarok1) bullseye; urgency=medium

  * New upstream version for Amarok Linux 3
  * update logos

 -- Amarok Linux <packages@amaroklinux.org>  Fri, 25 Mar 2022 10:36:33 -0300

amarokwelcome (2.5.2+amarok) bullseye; urgency=medium

  * New upstream release for Amarok Linux 3
  * Fork from mintwelcome for AmarokOS

 -- Amarok Linux <packages@amaroklinux.org>  Mon, 05 Jul 2021 16:06:20 -0300

mintwelcome (2.5.1) uma; urgency=medium

  [ NikoKrause ]
  * Add favorites applet to Cinnamon legacy layout (#45)

  [ Lurux ]
  * Update panel screenshots in layout selector - The screenshots for the "traditional" layout was still showcasing the old Mint-X theme despite the fact that both layouts now default to the Mint-Y theme, so I updated them accordingly. - The icons in the HiDPI screenshot of the "modern" layout were misaligned, so I also decided to redo these with the same panel arrangement.

  [ Shadofer ]
  * Remove add_codecs and pkg variables and provide a shorter way to determine whether or not media codecs should be installed. (#48)

 -- Clement Lefebvre <root@linuxmint.com>  Wed, 26 May 2021 12:44:38 +0100

mintwelcome (2.5.0) ulyssa; urgency=medium

  * Revert "Run mintdrivers without admin rights"

 -- Clement Lefebvre <root@linuxmint.com>  Thu, 10 Dec 2020 11:30:02 +0000

mintwelcome (2.4.9) ulyssa; urgency=medium

  * Add favorites applet to Cinnamon modern layout

 -- Clement Lefebvre <root@linuxmint.com>  Tue, 08 Dec 2020 13:05:48 +0000

mintwelcome (2.4.8) ulyssa; urgency=medium

  * Run mintdrivers without admin rights
  * l10n: Generate files

 -- Clement Lefebvre <root@linuxmint.com>  Mon, 30 Nov 2020 15:27:28 +0000

mintwelcome (2.4.7) ulyana; urgency=medium

  * Fix icons in Mint 20
  * Cinnamon: Bring back panel layout selection
  * l10n: Update POT

 -- Clement Lefebvre <root@linuxmint.com>  Thu, 18 Jun 2020 12:29:36 +0100

mintwelcome (2.4.6) ulyana; urgency=medium

  * Fix brown and green pictures

 -- Clement Lefebvre <root@linuxmint.com>  Sat, 06 Jun 2020 12:16:07 +0100

mintwelcome (2.4.5) ulyana; urgency=medium

  * Go back to original colors

 -- Clement Lefebvre <root@linuxmint.com>  Sat, 06 Jun 2020 12:03:31 +0100

mintwelcome (2.4.4) ulyana; urgency=medium

  * Colors: Add HiDPI support

 -- Clement Lefebvre <root@linuxmint.com>  Fri, 24 Apr 2020 15:17:15 +0100

mintwelcome (2.4.3) ulyana; urgency=medium

  [ Nikolay Nechaev ]
  * New mint logo in Welcome tab

  [ Clement Lefebvre ]
  * Switch icon to filled ring
  * First steps: Replace layout switcher with color chooser
  * l10n: Update POT
  * l10n: Generate desktop files

 -- Clement Lefebvre <root@linuxmint.com>  Fri, 24 Apr 2020 14:45:17 +0100

mintwelcome (2.4.2) tricia; urgency=medium

  [ Vincent Vermeulen ]
  * allow for window controls on the left

 -- Clement Lefebvre <root@linuxmint.com>  Wed, 15 Jan 2020 14:35:05 +0000

mintwelcome (2.4.1) tricia; urgency=medium

  * l10n: Fix msgid

 -- Clement Lefebvre <root@linuxmint.com>  Thu, 05 Dec 2019 16:12:16 +0000

mintwelcome (2.4.0) tricia; urgency=medium

  * Update app icon

 -- Clement Lefebvre <root@linuxmint.com>  Tue, 19 Nov 2019 10:44:06 +0000

mintwelcome (2.3.9) tricia; urgency=medium

  * Update cinnamon panel layouts
  * Update logo

 -- Clement Lefebvre <root@linuxmint.com>  Wed, 13 Nov 2019 11:46:32 +0100

mintwelcome (2.3.8) tina; urgency=medium

  * Fix generate_desktop_files

 -- Clement Lefebvre <root@linuxmint.com>  Sun, 30 Jun 2019 15:41:54 +0200

mintwelcome (2.3.7) tessa; urgency=medium

  * Desktop layouts: Remove user applet in modern layout

 -- Clement Lefebvre <root@linuxmint.com>  Mon, 17 Dec 2018 09:54:38 +0000

mintwelcome (2.3.6) tessa; urgency=medium

  [ Michael Webster ]
  * mintwelcome.py: Use surfaces to set the layout button images.

 -- Clement Lefebvre <root@linuxmint.com>  Thu, 06 Dec 2018 13:30:52 +0000

mintwelcome (2.3.5) tessa; urgency=medium

  * Load the non-Hidpi pics no matter what

 -- Clement Lefebvre <root@linuxmint.com>  Fri, 30 Nov 2018 14:47:02 +0000

mintwelcome (2.3.4) tessa; urgency=medium

  [ Isaac Carter ]
  * mintwelcome: Added Firewall Configuration to the Welcome app. (#36)

  [ Clement Lefebvre ]
  * Update some msgids
  * Add the ability to choose a Cinnamon desktop layout
  * l10n: Update POT
  * Cinnamon layout: Set the menu label
  * Panel layout: Add a little warning that customizations will be lost

 -- Clement Lefebvre <root@linuxmint.com>  Wed, 28 Nov 2018 11:31:49 +0000

mintwelcome (2.3.3) tara; urgency=medium

  * Remove "new features" section in LMDE

 -- Clement Lefebvre <root@linuxmint.com>  Sun, 05 Aug 2018 10:35:49 +0200

mintwelcome (2.3.2) tara; urgency=medium

  * First steps: Don't show mintdrivers button if mintdrivers isn't installed

 -- Clement Lefebvre <root@linuxmint.com>  Thu, 12 Jul 2018 12:03:33 +0200

mintwelcome (2.3.1) tara; urgency=medium

  [ NikoKrause ]
  * Translate Title "Welcome" (#34)

 -- Clement Lefebvre <root@linuxmint.com>  Wed, 30 May 2018 18:04:46 +0100

mintwelcome (2.3.0) tara; urgency=medium

  [ Butterfly ]
  *  Fix getting translations 2 (#33)

  [ Clement Lefebvre ]
  * l10n: Update files

 -- Clement Lefebvre <root@linuxmint.com>  Tue, 22 May 2018 13:21:53 +0100

mintwelcome (2.2.9) tara; urgency=medium

  [ Clement Lefebvre ]
  * Generate files and remove obsolete files

  [ Butterfly ]
  * Fix getting translations (#32)

 -- Clement Lefebvre <root@linuxmint.com>  Mon, 21 May 2018 12:46:05 +0100

mintwelcome (2.2.8) tara; urgency=medium

  * Fix icon in autostart file
  * Finish implementation of the UI revamp


usr/lib/amarok/amarokwelcome/amarokwelcome.py
#!/usr/bin/python3
import apt
import gettext
import gi
import os
import platform
import subprocess
import locale
import cairo
gi.require_version("Gtk", "3.0")
from gi.repository import Gtk, Gio, Gdk, GdkPixbuf

NORUN_FLAG = os.path.expanduser("~/.amarok/amarokwelcome/norun.flag")

# i18n
gettext.install("amarokwelcome", "/usr/share/amarok/locale")
from locale import gettext as _
locale.bindtextdomain("amarokwelcome", "/usr/share/amarok/locale")
locale.textdomain("amarokwelcome")

LAYOUT_STYLE_LEGACY, LAYOUT_STYLE_NEW = range(2)

class SidebarRow(Gtk.ListBoxRow):

    def __init__(self, page_widget, page_name, icon_name):
        Gtk.ListBoxRow.__init__(self)
        self.page_widget = page_widget
        box = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=6)
        box.set_border_width(6)
        image = Gtk.Image.new_from_icon_name(icon_name, Gtk.IconSize.BUTTON)
        box.pack_start(image, False, False, 0)
        label = Gtk.Label()
        label.set_text(page_name)
        box.pack_start(label, False, False, 0)
        self.add(box)

class amarokWelcome:

    def __init__(self):
        builder = Gtk.Builder()
        builder.set_translation_domain("amarokwelcome")
        builder.add_from_file('/usr/share/amarok/amarokwelcome/amarokwelcome.ui')

        window = builder.get_object("main_window")
        window.set_icon_name("amarokwelcome")
        window.set_position(Gtk.WindowPosition.CENTER)
        window.connect("destroy", Gtk.main_quit)

        with open("/etc/amarok/info") as f:
            config = dict([line.strip().split("=") for line in f])
        codename = config['CODENAME'].capitalize()
        edition = config['EDITION'].replace('"', '')
        release = config['RELEASE']
        desktop = config['DESKTOP']
#        release_notes = config['RELEASE_NOTES_URL']
#        new_features = config['NEW_FEATURES_URL']
        architecture = "64-bit"
        if platform.machine() != "x86_64":
            architecture = "32-bit"

        # distro-specific
        dist_name = "Amarok Linux"
        if os.path.exists("/usr/share/doc/debian-system-adjustments/copyright"):
            dist_name = "Amarok"

        # Setup the labels in the Mint badge
        builder.get_object("label_version").set_text("%s %s" % (dist_name, release))
        builder.get_object("label_edition").set_text("%s %s" % (edition, architecture))

        # Setup the main stack
        self.stack = Gtk.Stack()
        builder.get_object("center_box").pack_start(self.stack, True, True, 0)
        self.stack.set_transition_type(Gtk.StackTransitionType.CROSSFADE)
        self.stack.set_transition_duration(150)

        # Action buttons
        builder.get_object("button_forums").connect("clicked", self.visit, "https://amaroklinux.forumeiros.com")
        builder.get_object("button_contribute").connect("clicked", self.visit, "https://amaroklinux.forumeiros.com")
        builder.get_object("button_update").connect("clicked", self.launch, "gpk-update-viewer")
        builder.get_object("button_amaroksoftware").connect("clicked", self.launch, "amarok-software")
        builder.get_object("button_timeshift").connect("clicked", self.pkexec, "timeshift-gtk")
        builder.get_object("button_gufw").connect("clicked", self.launch, "gufw")
        builder.get_object("button_layout_legacy").connect("clicked", self.on_button_layout_clicked, LAYOUT_STYLE_LEGACY)
        builder.get_object("button_layout_new").connect("clicked", self.on_button_layout_clicked, LAYOUT_STYLE_NEW)

        # Settings button depends on DE
        de_is_cinnamon = False
        self.theme = None
        if os.getenv("XDG_CURRENT_DESKTOP") in ["Cinnamon", "X-Cinnamon"]:
            builder.get_object("button_settings").connect("clicked", self.launch, "cinnamon-settings")
            de_is_cinnamon = True
            self.theme = Gio.Settings(schema="org.cinnamon.desktop.interface").get_string("gtk-theme")
        elif os.getenv("XDG_CURRENT_DESKTOP") == "MATE":
            builder.get_object("button_settings").connect("clicked", self.launch, "mate-control-center")
        elif os.getenv("XDG_CURRENT_DESKTOP") == "XFCE":
            builder.get_object("button_settings").connect("clicked", self.launch, "xfce4-settings-manager")
        elif os.getenv("XDG_CURRENT_DESKTOP") == "BUDGIE":
            builder.get_object("button_settings").connect("clicked", self.launch, "gnome-control-center")
        elif os.getenv("XDG_CURRENT_DESKTOP") == "GNOME":
            builder.get_object("button_settings").connect("clicked", self.launch, "gnome-control-center")
        elif os.getenv("XDG_CURRENT_DESKTOP") == "KDE":
            builder.get_object("button_settings").connect("clicked", self.launch, "systemsettings")
        else:
            # Hide settings
            builder.get_object("box_first_steps").remove(builder.get_object("box_settings"))

        # Hide Cinnamon layout settings in other DEs
        if not de_is_cinnamon:
            builder.get_object("box_first_steps").remove(builder.get_object("box_cinnamon"))

        # Construct the stack switcher
        list_box = builder.get_object("list_navigation")

        page = builder.get_object("page_home")
        self.stack.add_named(page, "page_home")
        list_box.add(SidebarRow(page, _("Welcome"), "go-home-symbolic"))
        self.stack.set_visible_child(page)

        page = builder.get_object("page_first_steps")
        self.stack.add_named(page, "page_first_steps")
        list_box.add(SidebarRow(page, _("First Steps"), "dialog-information-symbolic"))

        page = builder.get_object("page_help")
        self.stack.add_named(page, "page_help")
        list_box.add(SidebarRow(page, _("Help"), "help-browser-symbolic"))

        page = builder.get_object("page_contribute")
        self.stack.add_named(page, "page_contribute")
        list_box.add(SidebarRow(page, _("Contribute"), "starred-symbolic"))

        list_box.connect("row-activated", self.sidebar_row_selected_cb)

        # Construct the bottom toolbar
        box = builder.get_object("toolbar_bottom")
        checkbox = Gtk.CheckButton()
        checkbox.set_label(_("Show this dialog at startup"))
        if not os.path.exists(NORUN_FLAG):
            checkbox.set_active(True)
        checkbox.connect("toggled", self.on_button_toggled)
        box.pack_end(checkbox)

        scale = window.get_scale_factor()

        self.color = "green"
        self.dark_mode = False

        # Use HIDPI pictures if appropriate
        if scale == 1:
            surface = self.surface_for_path("/usr/share/amarok/amarokwelcome/legacy.png", scale)
            builder.get_object("img_legacy").set_from_surface(surface)
            surface = self.surface_for_path("/usr/share/amarok/amarokwelcome/modern.png", scale)
            builder.get_object("img_modern").set_from_surface(surface)
        else:
            surface = self.surface_for_path("/usr/share/amarok/amarokwelcome/legacy-hidpi.png", scale)
            builder.get_object("img_legacy").set_from_surface(surface)
            surface = self.surface_for_path("/usr/share/amarok/amarokwelcome/modern-hidpi.png", scale)
            builder.get_object("img_modern").set_from_surface(surface)

        window.set_default_size(800, 500)
        window.show_all()

    def surface_for_path(self, path, scale):
        pixbuf = GdkPixbuf.Pixbuf.new_from_file(path)

        return Gdk.cairo_surface_create_from_pixbuf(pixbuf, scale)

    def sidebar_row_selected_cb(self, list_box, row):
        self.stack.set_visible_child(row.page_widget)

    def on_button_toggled(self, button):
        if button.get_active():
            if os.path.exists(NORUN_FLAG):
                os.system("rm -rf %s" % NORUN_FLAG)
        else:
            os.system("mkdir -p ~/.amarok/amarokwelcome")
            os.system("touch %s" % NORUN_FLAG)

    def on_button_layout_clicked (self, button, style):

        applets_legacy = ['panel1:left:0:menu@cinnamon.org',
                          'panel1:left:1:show-desktop@cinnamon.org',
                          'panel1:left:2:panel-launchers@cinnamon.org',
                          'panel1:left:3:window-list@cinnamon.org',
                          'panel1:right:0:systray@cinnamon.org',
                          'panel1:right:1:xapp-status@cinnamon.org',
                          'panel1:right:2:keyboard@cinnamon.org',
                          'panel1:right:3:notifications@cinnamon.org',
                          'panel1:right:4:printers@cinnamon.org',
                          'panel1:right:5:removable-drives@cinnamon.org',
                          'panel1:right:6:favorites@cinnamon.org',
                          'panel1:right:7:user@cinnamon.org',
                          'panel1:right:8:network@cinnamon.org',
                          'panel1:right:9:sound@cinnamon.org',
                          'panel1:right:10:power@cinnamon.org',
                          'panel1:right:11:calendar@cinnamon.org']

        applets_new = ['panel1:left:0:menu@cinnamon.org',
                       'panel1:left:1:show-desktop@cinnamon.org',
                       'panel1:left:2:grouped-window-list@cinnamon.org',
                       'panel1:right:0:systray@cinnamon.org',
                       'panel1:right:1:xapp-status@cinnamon.org',
                       'panel1:right:2:notifications@cinnamon.org',
                       'panel1:right:3:printers@cinnamon.org',
                       'panel1:right:4:removable-drives@cinnamon.org',
                       'panel1:right:5:keyboard@cinnamon.org',
                       'panel1:right:6:favorites@cinnamon.org',
                       'panel1:right:7:network@cinnamon.org',
                       'panel1:right:8:sound@cinnamon.org',
                       'panel1:right:9:power@cinnamon.org',
                       'panel1:right:10:calendar@cinnamon.org']

        settings = Gio.Settings("org.cinnamon")
        settings.set_strv("panels-enabled", ['1:0:bottom'])

        applets = applets_new
        left_icon_size = 0
        center_icon_size = 0
        right_icon_size = 0
        if style == LAYOUT_STYLE_LEGACY:
            applets = applets_legacy
            panel_size = 27
            menu_label = "Menu"
        elif style == LAYOUT_STYLE_NEW:
            panel_size = 40
            right_icon_size = 24
            menu_label = ""

        settings.set_strv("panels-height", ['1:%s' % panel_size])
        settings.set_strv("enabled-applets", applets)
        settings.set_string("app-menu-label", menu_label)
        settings.set_string("panel-zone-icon-sizes", "[{\"panelId\": 1, \"left\": %s, \"center\": %s, \"right\": %s}]" % (left_icon_size, center_icon_size, right_icon_size))
        os.system("cinnamon --replace &")

    def on_dark_mode_changed(self, button, state):
        self.dark_mode = state
        self.change_color()

    def on_color_button_clicked(self, button, color):
        self.color = color
        self.change_color()

    def change_color(self):
        theme = "Adapta-Eta"
        wm_theme = "Adapta-Eta"
        cinnamon_theme = "Adapta-Nokto-Eta"
        if self.dark_mode:
            theme = "%s-Eta" % theme
            wm_theme = "Adapta-Nokto-Eta"
        if self.color != "green":
            theme = "%s-%s" % (theme, self.color.title())
            cinnamon_theme = "Adapta-Nokto-Eta-%s" % self.color.title()

        if os.getenv("XDG_CURRENT_DESKTOP") in ["Cinnamon", "X-Cinnamon"]:
            settings = Gio.Settings(schema="org.cinnamon.desktop.interface")
            settings.set_string("gtk-theme", theme)
            settings.set_string("icon-theme", theme)
            Gio.Settings(schema="org.cinnamon.desktop.wm.preferences").set_string("theme", wm_theme)
            Gio.Settings(schema="org.cinnamon.theme").set_string("name", cinnamon_theme)
        elif os.getenv("XDG_CURRENT_DESKTOP") == "MATE":
            settings = Gio.Settings(schema="org.mate.interface")
            settings.set_string("gtk-theme", theme)
            settings.set_string("icon-theme", theme)
            Gio.Settings(schema="org.mate.Marco.general").set_string("theme", wm_theme)
        elif os.getenv("XDG_CURRENT_DESKTOP") == "XFCE":
            subprocess.call(["xfconf-query", "-c", "xsettings", "-p", "/Net/ThemeName", "-s", theme])
            subprocess.call(["xfconf-query", "-c", "xsettings", "-p", "/Net/IconThemeName", "-s", theme])
            subprocess.call(["xfconf-query", "-c", "xfwm4", "-p", "/general/theme", "-s", theme])

    def visit(self, button, url):
        subprocess.Popen(["xdg-open", url])

    def launch(self, button, command):
        subprocess.Popen([command])

    def pkexec(self, button, command):
        subprocess.Popen(["pkexec", command])

if __name__ == "__main__":
    amarokWelcome()
    Gtk.main()